cirq/protocols/apply_unitary.py

```diff
@@ -227,27 +264,48 @@ def apply_unitary(unitary_value: Any,
     if func is not None:
         result = func(args)
         if result is not NotImplemented and result is not None:
+            if result is args.target_tensor:
+                return result
+            # If any entries of unitary_value's shape are less than the
+            # corresponding entries of args.qid_shape, don't trust func to
+            # copy the untouched entries from args.target_tensor into result.
+            op_shape = qid_shape_protocol.qid_shape(unitary_value,
+                                                    (2,) * len(args.axes))
+            if any(op_shape[i] < args.qid_shape[axis]
+                   for i, axis in enumerate(args.axes)):
+                # Copy extra entries from args.target_tensor into result.
+                for i in range(len(args.axes)):
+                    axis = args.axes[i]
+                    op_d = op_shape[i]
+                    axis_d = args.qid_shape[axis]
+                    subspace = linalg.slice_for_qubits_equal_to(
+                        (axis,), qureg_value_tuple=(slice(op_d, axis_d),))
+                    result[subspace] = args.target_tensor[subspace]
             return result
 
     # Fallback to using the object's _unitary_ matrix.
     matrix = unitary(unitary_value, None)
     if matrix is not None:
-        # Special case for single-qubit operations.
-        if matrix.shape == (2, 2):
-            zero = args.subspace_index(0)
-            one = args.subspace_index(1)
+        # Special case for single-qubit operations.(2,) * (2 * len(args.axes))
+        if len(args.axes) == 1:
```

Comment:
This can be specific to the `(2,)` shape, because the tradeoff depends very much on the size of the matrix and the defualt einsum thing might already be winning by 3x3.

Reply:
I didn't realize that.  I can change it to `if len(args.axes) == 1 and matrix.shape == (2, 2):`.